src/remote-desktop.c

```diff
@@ -1350,6 +1355,116 @@ handle_notify_touch_up (XdpDbusRemoteDesktop *object,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session is already connected");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  switch (remote_desktop_session->state)
+    {
+    case REMOTE_DESKTOP_SESSION_STATE_STARTED:
+      break;
+    case REMOTE_DESKTOP_SESSION_STATE_INIT:
+    case REMOTE_DESKTOP_SESSION_STATE_DEVICES_SELECTED:
+    case REMOTE_DESKTOP_SESSION_STATE_SELECTING_DEVICES:
+    case REMOTE_DESKTOP_SESSION_STATE_SOURCES_SELECTED:
+    case REMOTE_DESKTOP_SESSION_STATE_SELECTING_SOURCES:
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session is not ready");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    case REMOTE_DESKTOP_SESSION_STATE_STARTING:
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session is already started");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    case REMOTE_DESKTOP_SESSION_STATE_CLOSED:
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session is already closed");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  if (!xdp_filter_options (arg_options, &options_builder,
+                           remote_desktop_connect_to_eis_options,
+                           G_N_ELEMENTS (remote_desktop_connect_to_eis_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  if (!xdp_dbus_impl_remote_desktop_call_connect_to_eis_sync (impl,
+                                                             arg_session_handle,
+                                                             xdp_app_info_get_id (call->app_info),
+                                                             g_variant_builder_end (&options_builder),
+                                                             in_fd_list,
+                                                             &fd,
+                                                             &out_fd_list,
+                                                             NULL,
+                                                             &error))
+    {
+      g_warning ("Failed to ConnectToEIS: %s", error->message);
+      out_fd_list = g_unix_fd_list_new ();
+    }
+
+  /* FIXME: libreis_set_name() and friends */
+
+  remote_desktop_session->uses_eis = TRUE;
```

Comment:
Will it make sense to do this only if the backend call succeeded so that we can allow the existing/slower input injection implementation as the fallback?

Reply:
yes, good point, I'll add that to my list

Reply:
done in the lastest commit now.